python/ql/src/experimental/semmle/python/Concepts.qll

```diff
@@ -40,31 +40,52 @@ module SystemCommandExecution {
 }
 
 /**
- * A data-flow node that performs deserialization in a potentially unsafe way.
+ * A function that decodes data from a binary or textual format.
+ * Doing so should normally preserve taint, but it can also be a problem
+ * in itself, e.g. if it performs deserialization in a potentially unsafe way.
  *
  * Extend this class to refine existing API models. If you want to model new APIs,
- * extend `DeserializationSink::Range` instead.
+ * extend `UnmarshalingFunction::Range` instead.
  */
-class DeserializationSink extends DataFlow::Node {
-  DeserializationSink::Range self;
+class UnmarshalingFunction extends DataFlow::Node {
+  UnmarshalingFunction::Range self;
 
-  DeserializationSink() { this = self }
+  UnmarshalingFunction() { this = self }
 
-  /** Gets the argument that specifies the command to be executed. */
-  DataFlow::Node getData() { result = self.getData() }
+  /** Holds if this call is unsafe, e.g. if it may execute arbitrary code. */
+  predicate unsafe() { self.unsafe() }
```

Comment:
I was hoping that this could cover many ways to be unsafe, even if we currently only detect one.

Reply:
I'm not sure I understand how this would work. Lets say YAML is also vulnerable to billion laughs attack. How can you make a query for _out-of-memory / denial-of-service from deserialization that flags up when user-controlled data is decoded using vulnerable XML/YAML libraries, without also flagging up XML decoding as an _unsafe deserialization that can execute arbitrary code_? (My assumption here is that we want to have separate queries, since the impact of the two vulnerabilities is vastly different).

Reply:
Ah, I see now. You would want a predicate for each kind of unsafe. So `mayExecute`, `mayBlowUp`, etc.. And I guess that in many cases, this cannot be determined from the call alone and we would need a separate analysis to conclude e.g. that sufficient limits have not been set.

Reply:
> And I guess that in many cases, this cannot be determined from the call alone and we would need a separate analysis to conclude e.g. that sufficient limits have not been set.

That might actually be true. Time will tell :wink:

Reply:
@yoff besides the "SafeLoader" matching only on name, I think renaming this predicate is all that is left before I will approve :stuck_out_tongue: